File: src/main/java/com/github/stars/controller/CloneController.java
package com.github.stars.controller;

import com.github.stars.entity.CloneTask;
import com.github.stars.entity.GithubRepo;
import com.github.stars.service.CloneService;
import com.github.stars.service.GithubRepoService;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.annotation.Resource;
import java.io.File;
import java.nio.charset.StandardCharsets;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

/**
 * 批量 Clone - 脚本生成 + 实际执行
 */
@RestController
@RequestMapping("/api/clone")
public class CloneController {

    @Resource
    private GithubRepoService githubRepoService;

    @Resource
    private CloneService cloneService;

    /** 默认克隆深度：1=浅克隆 */
    private static final int DEFAULT_CLONE_DEPTH = 1;
    /** 默认最大仓库大小(MB)，0=不限制 */
    private static final int DEFAULT_MAX_REPO_SIZE_MB = 500;

    @GetMapping("/config")
    public Map<String, Object> getCloneConfig() {
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("success", true);
        result.put("baseDirectory", cloneService.getBaseDirectory());
        result.put("subdirectoryHistory", cloneService.getSubdirectoryHistory());
        result.put("lastSubdirectory", cloneService.getLastSubdirectory());
        result.put("hasActiveTask", cloneService.hasActiveTask());
        result.put("defaultCloneDepth", DEFAULT_CLONE_DEPTH);
        result.put("defaultMaxRepoSizeMb", DEFAULT_MAX_REPO_SIZE_MB);
        return result;
    }

    /**
     * 检查目标磁盘空间
     */
    @GetMapping("/disk-space")
    public Map<String, Object> checkDiskSpace(
            @RequestParam(value = "subDirectory", defaultValue = "") String subDirectory,
            @RequestParam(value = "repoCount", defaultValue = "50") int repoCount) {
        Map<String, Object> result = new LinkedHashMap<>();
        Map<String, Object> diskInfo = cloneService.checkDiskSpace(subDirectory, repoCount);
        result.put("success", true);
        result.putAll(diskInfo);
        return result;
    }

    @PostMapping("/start")
    public Map<String, Object> startClone(
            @RequestParam(value = "keyword", defaultValue = "") String keyword,
            @RequestParam(value = "language", defaultValue = "") String language,
            @RequestParam(value = "categoryIds", defaultValue = "") String categoryIds,
            @RequestParam(value = "maxCount", defaultValue = "50") int maxCount,
            @RequestParam(value = "subDirectory", defaultValue = "") String subDirectory,
            @RequestParam(value = "dateField", defaultValue = "") String dateField,
            @RequestParam(value = "startDate", defaultValue = "") String startDate,
            @RequestParam(value = "endDate", defaultValue = "") String endDate,
            @RequestParam(value = "sortBy", defaultValue = "starred_at") String sortBy,
            @RequestParam(value = "sortOrder", defaultValue = "desc") String sortOrder,
            @RequestParam(value = "concurrency", defaultValue = "5") int concurrency,
            @RequestParam(value = "cloneDepth", defaultValue = "1") int cloneDepth,
            @RequestParam(value = "maxRepoSizeMb", defaultValue = "500") int maxRepoSizeMb) {
        Map<String, Object> result = new LinkedHashMap<>();
        try {
            String taskId = cloneService.startBatchClone(keyword, language, categoryIds, maxCount, subDirectory,
                    dateField, startDate, endDate, sortBy, sortOrder, concurrency,
                    cloneDepth, maxRepoSizeMb);
            result.put("success", true);
            result.put("taskId", taskId);
            result.put("targetDirectory", cloneService.resolveCloneDirectory(subDirectory).getAbsolutePath());
            result.put("message", String.format("Clone 任务已启动 (%d并发, depth=%d, maxSize=%dMB)",
                    concurrency, cloneDepth, maxRepoSizeMb));
        } catch (IllegalArgumentException | IllegalStateException e) {
            result.put("success", false);
            result.put("message", e.getMessage());
        }
        return result;
    }

    @GetMapping("/task/{taskId}")
    public Map<String, Object> getTaskProgress(@PathVariable String taskId) {
        Map<String, Object> result = new LinkedHashMap<>();
        CloneTask task = cloneService.getTask(taskId);
        if (task == null) {
            result.put("success", false);
            result.put("message", "任务不存在");
            return result;
        }
        result.put("success", true);
        result.put("taskId", task.getTaskId());
        result.put("status", task.getStatus());
        result.put("errorMessage", task.getErrorMessage());
        result.put("totalRepos", task.getTotalRepos());
        result.put("completedRepos", task.getCompletedRepos());
        result.put("failedRepos", task.getFailedRepos());
        result.put("skippedRepos", task.getSkippedRepos());
        result.put("cloneDepth", task.getCloneDepth());
        result.put("maxRepoSizeMb", task.getMaxRepoSizeMb());
        result.put("cancelled", task.getCancelled());
        result.put("results", task.getResults() != null ? task.getResults() : Collections.emptyList());
        return result;
    }

    /**
     * 取消正在运行的任务
     */
    @PostMapping("/task/{taskId}/cancel")
    public Map<String, Object> cancelTask(@PathVariable String taskId) {
        Map<String, Object> result = new LinkedHashMap<>();
        boolean cancelled = cloneService.cancelTask(taskId);
        result.put("success", cancelled);
        result.put("message", cancelled ? "任务已取消" : "无法取消（任务不存在或已完成）");
        return result;
    }

    @GetMapping("/script")
    public ResponseEntity<byte[]> generateScript(
            @RequestParam(value = "osType", defaultValue = "windows") String osType,
            @RequestParam(value = "keyword", defaultValue = "") String keyword,
            @RequestParam(value = "language", defaultValue = "") String language,
            @RequestParam(value = "categoryIds", defaultValue = "") String categoryIds,
            @RequestParam(value = "maxCount", defaultValue = "50") int maxCount,
            @RequestParam(value = "subDirectory", defaultValue = "") String subDirectory,
            @RequestParam(value = "cloneDepth", defaultValue = "1") int cloneDepth) {

        List<GithubRepo> repos = githubRepoService.findPage(1, maxCount, keyword, language,
                "starred_at", "desc", null, null, null, categoryIds).getRecords();

        File cloneDirFile = cloneService.resolveCloneDirectory(subDirectory);
        String cloneDir = cloneDirFile.getAbsolutePath();

        String depthFlag = cloneDepth > 0 ? " --depth " + cloneDepth : "";

        StringBuilder script = new StringBuilder();

        if ("windows".equals(osType)) {
            script.append("# GitHub Stars 批量 Clone 脚本\n");
            script.append("# 生成时间: ").append(LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"))).append("\n");
            script.append("# 项目数量: ").append(repos.size()).append("\n");
            script.append("# 克隆深度: ").append(cloneDepth > 0 ? String.valueOf(cloneDepth) : "完整克隆").append("\n\n");
            script.append("$cloneDir = \"").append(cloneDir).append("\"\n");
            script.append("if (-not (Test-Path $cloneDir)) { New-Item -ItemType Directory -Path $cloneDir -Force | Out-Null }\n");
            script.append("Set-Location $cloneDir\n\n");

            for (GithubRepo repo : repos) {
                script.append("# ").append(repo.getFullName());
                if (repo.getDescriptionCn() != null) script.append(" - ").append(repo.getDescriptionCn());
                else if (repo.getDescription() != null) script.append(" - ").append(repo.getDescription());
                script.append("\n");
                script.append("if (Test-Path \"").append(repo.getRepoName()).append("\") {\n");
                script.append("  Write-Host \"[SKIP] ").append(repo.getRepoName()).append(" 已存在\"\n");
                script.append("} else {\n");
                script.append("  git clone").append(depthFlag).append(" ").append(cloneService.buildCloneUrl(repo.getHtmlUrl())).append("\n");
                script.append("}\n\n");
            }
            script.append("Write-Host \"Done! Cloned into $cloneDir\"\n");
        } else {
            script.append("#!/bin/bash\n");
            script.append("# GitHub Stars 批量 Clone 脚本\n");
            script.append("# 生成时间: ").append(LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"))).append("\n");
            script.append("# 项目数量: ").append(repos.size()).append("\n");
            script.append("# 克隆深度: ").append(cloneDepth > 0 ? String.valueOf(cloneDepth) : "完整克隆").append("\n\n");
            script.append("CLONE_DIR=\"").append(cloneDir).append("\"\n");
            script.append("mkdir -p \"$CLONE_DIR\"\n");
            script.append("cd \"$CLONE_DIR\" || exit\n\n");

            for (GithubRepo repo : repos) {
                script.append("# ").append(repo.getFullName());
                if (repo.getDescriptionCn() != null) script.append(" - ").append(repo.getDescriptionCn());
                else if (repo.getDescription() != null) script.append(" - ").append(repo.getDescription());
                script.append("\n");
                script.append("if [ -d \"").append(repo.getRepoName()).append("\" ]; then\n");
                script.append("  echo \"[SKIP] ").append(repo.getRepoName()).append(" already exists\"\n");
                script.append("else\n");
                script.append("  git clone").append(depthFlag).append(" ").append(cloneService.buildCloneUrl(repo.getHtmlUrl())).append("\n");
                script.append("fi\n\n");
            }
            script.append("echo \"Done! Cloned into $CLONE_DIR\"\n");
        }

        byte[] bytes = script.toString().getBytes(StandardCharsets.UTF_8);
        String ext = "windows".equals(osType) ? "ps1" : "sh";
        String filename = "github_stars_clone_" + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss")) + "." + ext;
        String encodedFilename;
        try {
            encodedFilename = java.net.URLEncoder.encode(filename, "UTF-8");
        } catch (java.io.UnsupportedEncodingException e) {
            encodedFilename = filename;
        }

        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename*=UTF-8''" + encodedFilename)
                .contentType(MediaType.TEXT_PLAIN)
                .contentLength(bytes.length)
                .body(bytes);
    }

    @ExceptionHandler({IllegalArgumentException.class, IllegalStateException.class})
    public ResponseEntity<Map<String, Object>> handleCloneError(RuntimeException e) {
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("success", false);
        result.put("message", e.getMessage());
        return ResponseEntity.badRequest().body(result);
    }
}

File: backend-nestjs/src/clone/controllers/clone.controller.ts
import { Controller, Get, Post, Delete, Param, Query, Body, Res, Logger } from '@nestjs/common';
import type { Response } from 'express';
import { CloneService } from '../services/clone.service';
import { CloneTaskService } from '../services/clone-task.service';

@Controller()
export class CloneController {
    private readonly logger = new Logger(CloneController.name);

    constructor(
        private readonly cloneService: CloneService,
        private readonly taskService: CloneTaskService,
    ) {}

    /**
     * 获取克隆配置信息
     *
     * @returns 克隆配置，包含基础目录、子目录历史等
     */
    @Get('api/clone/config')
    async config() {
        return this.cloneService.getCloneConfig();
    }

    /**
     * 检查磁盘空间是否足够进行克隆
     *
     * @param q.subDirectory 子目录路径
     * @param q.repoCount 预计克隆的仓库数量
     * @returns 磁盘空间检查结果
     */
    @Get('api/clone/disk-space')
    async diskSpace(@Query() q: any) {
        return this.cloneService.checkDiskSpace(q.subDirectory || '', parseInt(q.repoCount) || 50);
    }

    /**
     * 启动批量克隆任务
     *
     * @param q 查询参数，包含筛选条件、并发数、克隆深度等
     * @returns 任务创建结果，包含 taskId
     */
    @Post('api/clone/start')
    async start(@Query() q: any) {
        this.logger.log('接收到批量克隆启动请求');
        return this.cloneService.startBatchClone({
            keyword: q.keyword || '',
            language: q.language || '',
            categoryIds: q.categoryIds || '',
            maxCount: parseInt(q.maxCount) || 50,
            subDirectory: q.subDirectory || '',
            dateField: q.dateField || '',
            startDate: q.startDate || '',
            endDate: q.endDate || '',
            sortBy: q.sortBy || 'starred_at',
            sortOrder: q.sortOrder || 'desc',
            concurrency: parseInt(q.concurrency) || 5,
            cloneDepth: parseInt(q.cloneDepth) || 1,
            maxRepoSizeMb: parseInt(q.maxRepoSizeMb) || 500,
        });
    }

    /**
     * 查询单个克隆任务状态，包含克隆结果列表
     *
     * @param taskId 任务 ID
     * @returns 任务状态及克隆结果
     */
    @Get('api/clone/task/:taskId')
    async task(@Param('taskId') taskId: string) {
        const t = await this.cloneService.getTask(taskId);
        if (!t) return { success: false, message: '任务不存在' };
        return {
            success: true,
            taskId: t.taskId,
            status: t.status,
            totalRepos: t.totalRepos,
            completedRepos: t.completedRepos,
            failedRepos: t.failedRepos,
            skippedRepos: t.skippedRepos,
            cancelled: t.cancelled,
            results: (t as any).items?.map((i: any) => ({ fullName: i.fullName, status: i.status, message: i.message })),
        };
    }

    /**
     * 取消正在运行或等待中的克隆任务
     *
     * @param taskId 任务 ID
     * @returns 取消结果
     */
    @Post('api/clone/task/:taskId/cancel')
    async cancel(@Param('taskId') taskId: string) {
        this.logger.log('接收到取消克隆任务请求: taskId=' + taskId);
        const ok = await this.cloneService.cancelTask(taskId);
        return { success: ok, message: ok ? '任务已取消' : '无法取消' };
    }

    /**
     * 分页查询克隆任务列表
     *
     * @param q.page 页码
     * @param q.size 每页数量
     * @returns 任务分页数据
     */
    @Get('api/clone/tasks')
    async tasks(@Query() q: any) {
        const result = await this.taskService.getTaskPage(parseInt(q.page) || 1, parseInt(q.size) || 20);
        return { success: true, ...result };
    }

    /**
     * 查询克隆任务详情（含分页任务项）
     *
     * @param taskId 任务 ID
     * @param q.page 页码
     * @param q.size 每页数量
     * @param q.status 可选的状态筛选
     * @returns 任务详情及分页任务项
     */
    @Get('api/clone/tasks/:taskId')
    async taskDetail(@Param('taskId') taskId: string, @Query() q: any) {
        const detail = await this.taskService.getTaskDetail(taskId, parseInt(q.page) || 1, parseInt(q.size) || 100, q.status || '');
        if (!detail) return { success: false, message: '任务不存在' };
        return { success: true, ...detail };
    }

    /**
     * 删除克隆任务及其所有任务项
     *
     * @param taskId 任务 ID
     * @returns 删除结果
     */
    @Delete('api/clone/tasks/:taskId')
    async deleteTask(@Param('taskId') taskId: string) {
        this.logger.log('删除克隆任务: taskId=' + taskId);
        await this.taskService.deleteTaskByTaskId(taskId);
        return { success: true, message: '任务已删除' };
    }

    /**
     * 分页查询指定任务的任务项列表
     *
     * @param taskId 任务 ID
     * @param q.page 页码
     * @param q.size 每页数量
     * @param q.status 可选的状态筛选
     * @returns 任务项分页数据
     */
    @Get('api/clone/tasks/:taskId/items')
    async taskItems(@Param('taskId') taskId: string, @Query() q: any) {
        const result = await this.taskService.getItemsByTaskId(taskId, parseInt(q.page) || 1, parseInt(q.size) || 100, q.status || '');
        return { success: true, ...result };
    }

    /**
     * 重试所有存在失败项的任务
     *
     * @returns 重试结果汇总
     */
    @Post('api/clone/tasks/retry-all')
    async retryAll() {
        try {
            const ids = await this.taskService.getTaskIdsWithFailedItems();
            if (!ids.length) return { success: false, message: '没有需要重试的任务' };
            this.logger.log('批量重试克隆失败项: 任务数=' + ids.length);
            let totalRetried = 0;
            for (const id of ids) {
                const r = await this.cloneService.retryFailedClones(id);
                if (r.success) totalRetried += (r as any).retryCount || 0;
            }
            return { success: true, message: `处理了 ${ids.length} 个任务，共重试 ${totalRetried} 个失败项` };
        } catch (e) {
            this.logger.error('批量重试克隆失败项异常: ' + (e instanceof Error ? e.message : String(e)));
            return { success: false, message: e instanceof Error ? e.message : String(e) };
        }
    }

    /**
     * 重试指定任务的失败项
     *
     * @param taskId 任务 ID
     * @returns 重试结果
     */
    @Post('api/clone/tasks/:taskId/retry')
    async retryTask(@Param('taskId') taskId: string) {
        this.logger.log('重试任务失败项: taskId=' + taskId);
        return this.cloneService.retryFailedClones(taskId);
    }

    /**
     * 切换任务的置顶状态
     *
     * @param taskId 任务 ID
     * @returns 置顶切换结果
     */
    @Post('api/clone/tasks/:taskId/pin')
    async pinTask(@Param('taskId') taskId: string) {
        try {
            const pinned = await this.taskService.togglePin(taskId);
            return { success: true, pinned, message: pinned ? '已置顶' : '已取消置顶' };
        } catch (e) {
            this.logger.error('任务置顶切换失败: taskId=' + taskId + ', 错误=' + (e instanceof Error ? e.message : String(e)));
            return { success: false, message: e instanceof Error ? e.message : String(e) };
        }
    }

    /**
     * 生成可下载的克隆脚本文件（PowerShell 或 Bash）
     *
     * @param q.osType 操作系统类型（windows/linux）
     * @param q 其他筛选参数
     * @param res Express 响应对象，用于直接输出文件下载
     */
    @Get('api/clone/script')
    async script(@Query() q: any, @Res() res: Response) {
        const script = await this.cloneService.generateCloneScript({
            osType: q.osType || 'windows',
            keyword: q.keyword || '',
            language: q.language || '',
            categoryIds: q.categoryIds || '',
            maxCount: parseInt(q.maxCount) || 50,
            subDirectory: q.subDirectory || '',
            cloneDepth: parseInt(q.cloneDepth) || 1,
        });
        const ext = q.osType === 'linux' ? 'sh' : 'ps1';
        res.set({
            'Content-Type': 'text/plain; charset=utf-8',
            'Content-Disposition': `attachment; filename*=UTF-8''${encodeURIComponent('clone-script.' + ext)}`,
        });
        res.send(script);
    }
}
